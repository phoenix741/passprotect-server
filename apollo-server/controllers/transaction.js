import { checkPermission } from '../utils/authentification'
import { getTransactions } from '../services/transaction'
import { pubsub, TRANSACTION_ADDED_TOPIC } from '../services/subscriptions'
import { withFilter } from 'graphql-subscriptions'
import moment from 'moment'

export const resolvers = {
  RootQuery: {
    async transactions (obj, { earliest }, { user }) {
      checkPermission(user)
      earliest = moment(earliest).toDate()

      return getTransactions(user, { earliest })
    }
  },

  RootSubscription: {
    transactionAdded: {
      subscribe: withFilter(() => pubsub.asyncIterator(TRANSACTION_ADDED_TOPIC), (payload, context) => {
        return payload.transactionAdded.user === context.user._id
      })
    }
  }
}
