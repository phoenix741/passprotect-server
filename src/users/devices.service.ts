import { Model } from 'mongoose';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { UserEntity } from './models/user.entity';
import { DeviceEntity } from './models/user.entity';
import { UsersService } from './users.service';

@Injectable()
export class DevicesService {
  constructor(private readonly userService: UsersService) {}

  async findAll(userId: string): Promise<DeviceEntity[]> {
    if (userId) {
      const user = await this.userService.findById(userId);
      return user.devices;
    }
    return [];
  }

  async verifyDevice(user: UserEntity, deviceId: string): Promise<UserEntity> {
    const isValid = user.devices.find(d => d.id === deviceId);
    if (!isValid) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }
    return user;
  }

  async registerDevice(userId: string, device: DeviceEntity): Promise<DeviceEntity> {
    const user = await this.userService.findById(userId);
    if (!user) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }

    const found = user.devices.find(d => d.id === device.id);
    if (!found) {
      user.devices.push(device);
      await user.save();
    }
    return found || device;
  }

  async removeDevice(userId: string, deviceId: string): Promise<void> {
    const user = await this.userService.findById(userId);
    if (!user) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }

    user.devices = user.devices.filter(d => d.id !== deviceId);
    await user.save();
  }
}
