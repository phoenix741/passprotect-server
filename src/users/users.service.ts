import { Model } from 'mongoose';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { UserToRegister } from './models/user.models';
import { UserEntity } from './models/user.entity';
import { CryptoService } from '../shared/services/crypto.service';
import { DeviceEntity } from '../users/models/user.entity';

@Injectable()
export class UsersService {
  constructor(private readonly cryptoService: CryptoService, @InjectModel('User') private readonly userModel: Model<UserEntity>) {}

  async findAll(): Promise<UserEntity[]> {
    return await this.userModel.find().exec();
  }

  async findById(id: string): Promise<UserEntity | null> {
    return id && (await this.userModel.findById(id.toLowerCase()));
  }

  async registerUser(user: UserToRegister): Promise<UserEntity> {
    user.password = await this.cryptoService.hashPassword(user.password);

    return await this.userModel.create(user);
  }

  async verifyPassword(user: UserEntity, password: string): Promise<UserEntity> {
    const isValid = await this.cryptoService.checkPassword(password, user.password);
    if (!isValid) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }
    return user;
  }
}
