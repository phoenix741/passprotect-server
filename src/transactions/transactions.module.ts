import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { TransactionSchema } from './schemas/transaction.schema';
import { TransactionsService } from './transactions.service';
import { SharedModule } from '../shared/shared.module';

@Module({
  imports: [
    SharedModule,
    MongooseModule.forFeature([{ name: 'Transaction', schema: TransactionSchema }]),
  ],
  providers: [TransactionsService],
  exports: [TransactionsService],
})
export class TransactionsModule {}
