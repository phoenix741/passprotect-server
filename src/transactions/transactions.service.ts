import { Model } from 'mongoose';
import { PubSub } from 'apollo-server-express';
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { TransactionTypeEnum } from '../shared/interfaces/transaction-type-enum.interface';
import { TransactionEntity } from './models/transaction.entity';
import { IFindAllTransactionParams } from './models/transaction.models';
import { LineEntity } from '../lines/models/line.entity';
import { createHash } from 'crypto';
import { ObjectID, Binary } from 'bson';

export const TRANSACTION_ADDED_TOPIC = 'transactionAdded';

@Injectable()
export class TransactionsService {
  public transactionPubSub = new PubSub();

  constructor(
    @InjectModel('Transaction')
    private readonly transactionModel: Model<TransactionEntity>,
  ) {}

  async findAll(userId: string, params: IFindAllTransactionParams = {}): Promise<TransactionEntity[]> {
    const filter: IFindAllTransactionDto = { user: userId };
    if (params.earliest) {
      filter.updatedAt = { $gte: params.earliest };
    }
    return await this.transactionModel
      .find(filter)
      .sort({ updatedAt: 1 })
      .exec();
  }

  async createTransaction(type: TransactionTypeEnum, before?: LineEntity, after?: LineEntity): Promise<TransactionEntity> {
    const base = after || before;
    if (!base) {
      throw new Error('Transaction: before or after should be defined');
    }

    const transaction: ICreateTransactionDto = {
      _id: new ObjectID(),
      type,
      line: base._id,
      user: base.user,
      before,
      after,
      sha512:
        after &&
        createHash('sha512')
          .update(after.encryption.content instanceof Binary ? after.encryption.content.buffer : after.encryption.content)
          .digest('hex'),
    };

    const createdTransaction = await this.transactionModel.create(transaction);
    this.transactionAdded(createdTransaction);
    return createdTransaction;
  }

  /**
   * Publish a new transaction to applications.
   *
   * @param {Object} transaction Transaction object
   */
  private transactionAdded(transaction: TransactionEntity) {
    this.transactionPubSub.publish(TRANSACTION_ADDED_TOPIC, {
      transactionAdded: transaction,
    });
  }
}

interface IFindAllTransactionDto {
  user: string;
  updatedAt?: {
    $gte: Date;
  };
}

interface ICreateTransactionDto {
  _id: ObjectID;
  type: TransactionTypeEnum;
  line: ObjectID;
  user: string;
  before: object;
  after: object;
  sha512: string;
}
