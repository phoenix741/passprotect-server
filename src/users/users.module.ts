import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { UserSchema } from './schemas/user.schema';
import { UsersService } from './users.service';
import { SharedModule } from '../shared/shared.module';
import { DevicesService } from './devices.service';

@Module({
  imports: [MongooseModule.forFeature([{ name: 'User', schema: UserSchema }]), SharedModule],
  providers: [UsersService, DevicesService],
  exports: [UsersService, DevicesService],
})
export class UsersModule {}
