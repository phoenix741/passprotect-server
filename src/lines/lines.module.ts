import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { LineSchema } from './schemas/line.schema';
import { LinesService } from './lines.service';
import { SharedModule } from '../shared/shared.module';
import { TransactionsModule } from '../transactions/transactions.module';

@Module({
  imports: [SharedModule, MongooseModule.forFeature([{ name: 'Line', schema: LineSchema }]), TransactionsModule],
  providers: [LinesService],
  exports: [LinesService],
})
export class LinesModule {}
