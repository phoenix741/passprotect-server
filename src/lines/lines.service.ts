import { Model } from 'mongoose';
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { LineToCreate, LineToUpdate } from './models/line.models';
import { LineEntity } from './models/line.entity';
import { ObjectID } from 'bson';
import { TransactionsService } from '../transactions/transactions.service';
import { TransactionEntity } from '../transactions/models/transaction.entity';
import { TransactionTypeEnum } from '../shared/interfaces/transaction-type-enum.interface';

interface IFindAllFilter {
  user?: string;
}

@Injectable()
export class LinesService {
  constructor(
    @InjectModel('Line') private readonly lineModel: Model<LineEntity>,
    private readonly transactionService: TransactionsService,
  ) {}

  async findAll(userId: string): Promise<LineEntity[]> {
    if (userId) {
      const filter: IFindAllFilter = { user: userId };
      const sort = { group: 1, label: 1 };
      return await this.lineModel
        .find(filter)
        .sort(sort)
        .exec();
    }
    return [];
  }

  async findById(id: ObjectID) {
    return await this.lineModel.findById(id);
  }

  async createLine(line: LineToCreate): Promise<LineEntity> {
    const lineToCreate = new this.lineModel(line);
    const newLine = await lineToCreate.save();
    await this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      null,
      newLine,
    );
    return newLine;
  }

  async updateLine(
    originalLine: LineEntity,
    line: LineToUpdate,
  ): Promise<LineEntity> {
    const { _id, ...cleanLine } = line;

    const oldLine = new this.lineModel(originalLine.toObject());
    originalLine.set(cleanLine);
    const newLine = await originalLine.save();

    await this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      oldLine,
      newLine,
    );
    return newLine;
  }

  async removeLine(id: ObjectID): Promise<TransactionEntity> {
    const oldLine = await this.lineModel.findById(id);
    await this.lineModel.deleteOne({ _id: id });
    return this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      oldLine,
      null,
    );
  }

  async getGroups(userId: string): Promise<string[]> {
    const filter = { user: userId };
    return this.lineModel.distinct('group', filter);
  }
}
