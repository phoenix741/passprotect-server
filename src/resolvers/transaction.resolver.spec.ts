import { TransactionResolver } from './transaction.resolver';
import { PubSub } from 'graphql-subscriptions';
import { UsersService } from 'src/users/users.service';
import { LinesService } from 'src/lines/lines.service';
import { AuthorizationService } from 'src/shared/services/authorization.service';
import { TransactionsService } from 'src/transactions/transactions.service';
import { Test } from '@nestjs/testing';
import { TransactionEntity } from 'src/transactions/models/transaction.entity';
import { ObjectID } from 'bson';
import { TransactionTypeEnum } from 'src/shared/interfaces/transaction-type-enum.interface';
import { UserEntity } from 'src/users/models/user.entity';
import { LineTypeEnum } from 'src/shared/interfaces/line-type-enum.interface';

describe('TransactionResolver', () => {
  let transactionResolver: TransactionResolver;
  let userService;
  let linesService;
  let transactionService;
  let authorizationService;

  const user: UserEntity = {
    _id: 'username',
  } as any as UserEntity;
  const transaction: TransactionEntity = {
    _id: new ObjectID('5ce934e817517d7fb63cc425'),
    type: TransactionTypeEnum.line,
    line: new ObjectID('5ce934e817517d7fb63cc428'),
    user: 'usertransaction',
    before: {
      _id: new ObjectID('5ce934e817517d7fb63cc428'),
      type: LineTypeEnum.text,
      label: 'before',
    },
    after: {
      _id: new ObjectID('5ce934e817517d7fb63cc428'),
      type: LineTypeEnum.text,
      label: 'after',
    },
    sha512: 'sha512',
    createdAt: new Date('2019-05-25T12:28:24.941Z'),
    updatedAt: new Date('2019-05-25T12:28:24.941Z'),
    toObject() {
      return transaction;
    },
  } as TransactionEntity;

  beforeEach(async () => {
    userService = {
      findById: jest.fn(),
    };
    linesService = {
      findById: jest.fn(),
    };
    authorizationService = {
      checkPermission: jest.fn(),
    };
    transactionService = {
      findAll: jest.fn(),
      transactionPubSub: new PubSub(),
    };

    const module = await Test.createTestingModule({
      providers: [
        TransactionResolver,
        {
          provide: UsersService,
          useValue: userService,
        },
        {
          provide: LinesService,
          useValue: linesService,
        },
        {
          provide: TransactionsService,
          useValue: transactionService,
        },
        {
          provide: AuthorizationService,
          useValue: authorizationService,
        },
      ],
    })
    .compile();

    transactionResolver = module.get<TransactionResolver>(TransactionResolver);
  });

  describe('transactions', () => {
    it('success', async () => {
      transactionService.findAll.mockImplementation(() => [transaction]);
      expect(await transactionResolver.transactions(new Date('2019-05-25T12:28:24.941Z'), user)).toMatchSnapshot();
      expect(authorizationService.checkPermission).toHaveBeenCalledWith(user);
      expect(transactionService.findAll).toHaveBeenCalledWith('username', {earliest: new Date('2019-05-25T12:28:24.941Z')});
    });
  });
});
