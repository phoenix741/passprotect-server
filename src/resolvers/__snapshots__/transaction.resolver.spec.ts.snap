// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionResolver transactions success 1`] = `
Array [
  Object {
    "_id": "5ce934e817517d7fb63cc425",
    "after": Object {
      "_id": "5ce934e817517d7fb63cc428",
      "label": "after",
      "type": "text",
    },
    "before": Object {
      "_id": "5ce934e817517d7fb63cc428",
      "label": "before",
      "type": "text",
    },
    "createdAt": 2019-05-25T12:28:24.941Z,
    "line": "5ce934e817517d7fb63cc428",
    "sha512": "sha512",
    "toObject": [Function],
    "type": "line",
    "updatedAt": 2019-05-25T12:28:24.941Z,
    "user": "usertransaction",
  },
]
`;
